pkg/splunk/enterprise/clustermaster.go

```diff
@@ -0,0 +1,145 @@
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	splcommon "github.com/splunk/splunk-operator/pkg/splunk/common"
+	splctrl "github.com/splunk/splunk-operator/pkg/splunk/controller"
+)
+
+// ApplyClusterMaster reconciles the state of a Splunk Enterprise cluster master.
+func ApplyClusterMaster(client splcommon.ControllerClient, cr *enterprisev1.ClusterMaster) (reconcile.Result, error) {
+
+	// unless modified, reconcile for this object will be requeued after 5 seconds
+	result := reconcile.Result{
+		Requeue:      true,
+		RequeueAfter: time.Second * 5,
+	}
+	scopedLog := log.WithName("ApplyClusterMaster").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+
+	// validate and updates defaults for CR
+	err := validateClusterMasterSpec(cr)
+	if err != nil {
+		// To do: sgontla: later delete these listings. (for now just to test CSPL-320)
+		LogSmartStoreVolumes(cr.Status.SmartStore.VolList)
+		LogSmartStoreIndexes(cr.Status.SmartStore.IndexList)
+		return result, err
+	}
+
+	// updates status after function completes
+	cr.Status.Phase = splcommon.PhaseError
+	cr.Status.Selector = fmt.Sprintf("app.kubernetes.io/instance=splunk-%s-cluster-master", cr.GetName())
+	if !reflect.DeepEqual(cr.Status.SmartStore, cr.Spec.SmartStore) {
+		_, err := CreateSmartStoreConfigMap(client, cr, &cr.Spec.SmartStore)
+		if err != nil {
+			return result, err
+		}
+
+		// To do: sgontla: Do we need to update the status in K8 etcd immediately?
+		// Consider the case, where the Cluster master validates the config, and
+		// generates config map, but fails later in the flow to complete its
+		// stateful set. Meanwhile, all the indexer cluster sites notices new
+		// config map, and keeps resetting the indexers. This is not problem,
+		// as long as:
+		// 1. ApplyConfigMap avoids a CRUD update if there is no change to data,
+		// which is already happening today, so, this scenario shouldn't happen.
+		// 2. To Do: Is there anything additional to be done on indexer site?
+		cr.Status.SmartStore = cr.Spec.SmartStore
+	}
+
+	defer func() {
+		err = client.Status().Update(context.TODO(), cr)
+		if err != nil {
+			scopedLog.Error(err, "Status update failed")
+		}
+	}()
+
+	// check if deletion has been requested
+	if cr.ObjectMeta.DeletionTimestamp != nil {
+		terminating, err := splctrl.CheckForDeletion(cr, client)
+		if terminating && err != nil { // don't bother if no error, since it will just be removed immmediately after
+			cr.Status.Phase = splcommon.PhaseTerminating
+		} else {
+			result.Requeue = false
+		}
+		return result, err
+	}
+
+	// create or update general config resources
+	_, err = ApplySplunkConfig(client, cr, cr.Spec.CommonSplunkSpec, SplunkIndexer)
+	if err != nil {
+		return result, err
+	}
+
+	// create or update a headless service for indexer cluster
+	err = splctrl.ApplyService(client, getSplunkService(cr, &cr.Spec.CommonSplunkSpec, SplunkIndexer, true))
```

Comment:
Do we need to create SplunkIndexer service here? This can be done while creating an indexer

Reply:
+1 seems the two services should be created by IndexerCluster CRD versus ClusterMaster.

Kind of separate topic but related to this PR: Is "IndexerCluster" still the right name?

Reply:
This was deliberately done as part of multi-part approach. More discussion can be found here - https://github.com/splunk/splunk-operator/pull/99#discussion_r466094117.
PS: I raised the same question in that PR too, but I know now why it makes sense. :)

Reply:
oic forgot all about that.. so the cluster master's indexer service load balances across all indexers that are using the same cluster master? makes sense for regular (headed?) service, but not sure what the headless service would be for?

Reply:
hmm, ok I think we don't need it. Headless service on indexers should be enough.